vortex-serde/src/layouts/read/selection.rs

```diff
@@ -0,0 +1,164 @@
+                    }
+                    RowSelector::new(bitmap, 0, end)
+                })
+        })
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.values.is_empty()
+    }
+
+    pub fn begin(&self) -> usize {
+        self.begin
+    }
+
+    pub fn end(&self) -> usize {
+        self.end
+    }
+
+    pub fn len(&self) -> usize {
+        self.end - self.begin
+    }
+
+    pub fn intersect(mut self, other: &RowSelector) -> Self {
+        self.values.and_inplace(&other.values);
+        RowSelector::new(
+            self.values,
+            max(self.begin, other.begin),
+            min(self.end, other.end),
+        )
+    }
+
+    pub fn concatenate(mut self, other: &RowSelector) -> Self {
+        assert_eq!(
+            self.end, other.begin,
+            "Can only concatenate consecutive selectors"
+        );
+        self.values.or_inplace(&other.values);
+        RowSelector::new(
+            self.values,
+            min(self.begin, other.begin),
+            max(self.end, other.end),
+        )
+    }
+
+    pub fn filter_array(&self, array: impl AsRef<Array>) -> VortexResult<Option<Array>> {
+        if self.begin != 0 {
+            vortex_bail!("Cannot filter with selections that don't start at offset 0")
+        }
+
+        let true_count = self.values.cardinality();
+        if true_count == 0 {
+            return Ok(None);
+        }
+
+        let array = array.as_ref();
+
+        if true_count == array.len() as u64 {
+            return Ok(Some(array.clone()));
+        }
+
+        let bitset = self
+            .values
+            .to_bitset()
+            .ok_or_else(|| vortex_err!("Couldn't create bitset for RowSelection"))?;
+
+        let byte_length = self.len().div_ceil(8);
+        let mut buffer = MutableBuffer::with_capacity(byte_length);
+        buffer.extend_from_slice(bitset.as_slice());
+        if byte_length > bitset.size_in_bytes() {
+            buffer.extend_zeros(byte_length - bitset.size_in_bytes());
+        }
+        let predicate = BoolArray::try_new(
+            BooleanBuffer::new(buffer.into(), 0, self.len()),
+            Validity::NonNullable,
+        )?;
+        filter(array, predicate).map(Some)
+    }
+
+    pub fn offset(self, offset: i64) -> RowSelector {
+        if offset == 0 {
+            self
+        } else {
+            RowSelector::new(
+                self.values.add_offset(-offset),
+                if self.begin as i64 > offset {
+                    (self.begin as i64 - offset) as usize
+                } else {
+                    0
+                },
+                (self.end as i64 - offset) as usize,
+            )
+        }
+    }
+
+    pub fn advance(mut self, up_to_row: usize) -> Option<RowSelector> {
+        if up_to_row >= self.end {
+            None
+        } else {
+            self.values.remove_range(0..up_to_row as u32);
+            Some(RowSelector::new(self.values, self.begin, self.end))
```

Comment:
Should this not adjust `begin = min(self.begin, up_to_row)`?

Reply:
I think this could be implemented by just adjusting begin and then when you convert array to filter you remove everything before begin